### PR TITLE
ci: canary tests for direct/indirect dependents

### LINKED ISSUE #63: Add canary tests with `neostandard`-using modules
As eg. `fastify` has now migrated, https://github.com/fastify/fastify/pull/5509, we should add automated canary tests there as well, not just for modules using the old `standard` modules

### PR DESCRIPTION
**Pull request by voxpelli**
Fixes #63

Also contains a streamlining of the npm caches to ensure that they don't spiral out of control and eat up all of our cache space by growing infinitely larger over time

### CODE REVIEW

File: dependents-data/dependents.json
Comment (jerome-benoit) [inline dependents-data/dependents.json:3]: fastify?

Comment (voxpelli) [reply dependents-data/dependents.json:3]: Yes: https://github.com/fastify/fastify/pull/5509

File: dependents-data/dependents.json
Comment (voxpelli) [reply dependents-data/dependents.json]: Thanks!

**Comment by voxpelli** [comment]
I mistakenly pushed an attempt at fixing this directly to `main` rather than as a PR: https://github.com/neostandard/neostandard/commit/26b472b52b32d704f60f12bfb228692f2f74b2f6

**Comment by voxpelli** [comment]
@jerome-benoit After #86 is merged I'll release a new version and we can both update our projects to use the exported `eslint-plugin-n` rather than our own `eslint-plugin-n` in our configs, and it will unblock this

File: dependents-data/dependents.json
Comment (jerome-benoit) [inline dependents-data/dependents.json:4]: Is there a way to specify the directory to lint before bundling? 
Because the repo contains a lot of code not shipped in the built bundle as well as not the bundled code.

Comment (voxpelli) [reply dependents-data/dependents.json:4]: That would require a bit more of a complex setup in the GitHub Actions, doable maybe, but I would do it as a follow up then

Comment (jerome-benoit) [reply dependents-data/dependents.json:4]: On the repo side? like an added field to package.json? I would argue that the maintenance burden is pushed on the repo maintainers instead of neostandard ones, maybe with a notification mechanism to repo maintainers in case of failure. 

Without such configuration tunable, poolifier repo will never pass the current test. And I'm pretty sure other repos using bundler will also not pass it.

Comment (voxpelli) [reply dependents-data/dependents.json:4]: I would like to know more of how to support canary tests of them the best, shall we remove it from here for now and handle it in a follow up issue?

Comment (jerome-benoit) [reply dependents-data/dependents.json:4]: Removing poolifier repo until RFC design on how to support it is done is the best move for that PR.

Comment (jerome-benoit) [inline dependents-data/dependents.json:4]:
```suggestion
```

**Comment by voxpelli** [comment]
The two failures in my modules are because they contain committed lock files, need to investigate

File: .github/workflows/canary.yml
Comment (voxpelli) [inline .github/workflows/canary.yml:4]: When it all works, then move this to have it be triggered the same way that the compatibility tests are: From test workflow and only after linting and tests has passed

Comment (voxpelli) [reply .github/workflows/canary.yml]: Done ✔️

File: dependents-data/dependents.json
Comment (voxpelli) [reply dependents-data/dependents.json]: Done ✔️

File: .github/workflows/test.yml
Comment (jerome-benoit) [inline .github/workflows/test.yml:48]: I think that part can use a node version matrix to avoid code duplication. The condition might become a bit more complicated though.

Comment (voxpelli) [reply .github/workflows/test.yml:48]: I do that in some other repositories, but that involves splitting a string into an array of strings through a two job process, and in this case I felt it was cleaner to just repeat the test three times – especially the runs becomes cleaner as they have no parse step

(Would be kind of neat to be able to send array data to a reusable workflow, but as far as I'm aware only strings and booleans can be sent)

Comment (voxpelli) [reply .github/workflows/test.yml:48]: I guess another solution would be to pull them out into another reusable workflow and have that workflow use a matrix – so recursive reusable workflows, but that also feels a bit overkill

Comment (jerome-benoit) [reply .github/workflows/test.yml:48]:
```yaml
 test:
    name: Test on Node ${{ matrix.node }}
    uses: ./.github/workflows/reusable-npm-run.yml
    with:
      command: test-ci
      node-version: ${{ matrix.node }}
```

will not run it on all the node matrix?

**Comment by jerome-benoit** [comment]
Except nitpicking, I do not have a lot to say on that PR. I can approve it but I guess you need an approval from someone in the org team to be able to merge it.

**Review by jerome-benoit — APPROVED** [review]

File: .github/workflows/test.yml
Comment (voxpelli) [reply .github/workflows/test.yml:48]: I was 100% sure that wasn't possible, maybe they added that recently: https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-a-matrix-strategy-with-a-reusable-workflow

I'll fix!

Comment (voxpelli) [reply .github/workflows/test.yml]: Seems to work well 👍 Thanks for the suggestion @jerome-benoit!

Now I can remove the extra parsing step from all of my own reusable workflows https://github.com/voxpelli/ghatemplates/blob/0a20a5f68342d7c162637c56a856523592f4f982/.github/workflows/test.yml#L41-L51

**Comment by voxpelli** [comment]
> but I guess you need an approval from someone in the org team to be able to merge it.

I mostly need a sanity check, especially when it comes to things like these

**Pull request merged**
